File: C04_InnerTube/InnerTubeVB/SharedUtilities/Serializer.vb
Imports Microsoft.VisualBasic
Imports System.Xml.Serialization
Imports System.IO

Namespace SharedUtilities
	Public Class Serializer(Of T)

        Public Sub Serialize(ByVal value As T, ByVal path As String)
            'Tell the serializer what type we're serializing
            Dim xs As New XmlSerializer(GetType(T))

            Using write As TextWriter = New StreamWriter(path)
                xs.Serialize(write, value)
            End Using

        End Sub

		Public Function Deserialize(ByVal path As String) As T
			Dim xs As New XmlSerializer(GetType(T))

			Using read As TextReader = New StreamReader(path)
				Return CType(xs.Deserialize(read), T)
			End Using
		End Function
	End Class
End Namespace

File: C4_InnerTube/InnerTubeCS/SharedUtilities/Serializer.cs
using System.Xml.Serialization;
using System.IO;

namespace SharedUtilities
{
    public class Serializer<T>
    {
        public void Serialize(T value, string path)
        {
            //Tell the serializer what type we're serializing
            XmlSerializer xs = new XmlSerializer(typeof(T));

            using (TextWriter write = new StreamWriter(path))
            {
                xs.Serialize(write, value);
            }          

        }

        public T Deserialize(string path)
        {
            XmlSerializer xs = new XmlSerializer(typeof(T));
            
            using (TextReader read = new StreamReader(path))
            {
                return (T)xs.Deserialize(read);
            }
        }
    }
}
